File: ops/remote-browser/src/tls.mjs
'use strict'

import { execFileSync } from 'node:child_process'
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'

export function userdata() {
  return path.join(os.homedir(), '.zedcafe-remote-browser')
}

export function certpaths() {
  const dir = path.join(userdata(), 'tls')
  fs.mkdirSync(dir, { recursive: true })
  return {
    dir,
    key: path.join(dir, 'server.key'),
    cert: path.join(dir, 'server.crt'),
  }
}

export function ensureservercerts() {
  const paths = certpaths()
  if (fs.existsSync(paths.key) && fs.existsSync(paths.cert)) {
    return paths
  }
  execFileSync(
    'openssl',
    [
      'req',
      '-x509',
      '-newkey',
      'rsa:2048',
      '-sha256',
      '-days',
      '3650',
      '-nodes',
      '-keyout',
      paths.key,
      '-out',
      paths.cert,
      '-subj',
      '/CN=127.0.0.1',
      '-addext',
      'subjectAltName=IP:127.0.0.1,DNS:localhost',
    ],
    { stdio: 'pipe' },
  )
  return paths
}

export function installtrust(certpath) {
  if (process.platform === 'darwin') {
    try {
      execFileSync(
        'security',
        [
          'add-trusted-cert',
          '-d',
          '-r',
          'trustRoot',
          '-k',
          `${process.env.HOME}/Library/Keychains/login.keychain-db`,
          certpath,
        ],
        { stdio: 'pipe' },
      )
      return true
    } catch (err) {
      console.warn('macos trust install failed', err.message)
      return false
    }
  }
  if (process.platform === 'win32') {
    try {
      execFileSync('certutil', ['-addstore', '-user', 'Root', certpath], {
        stdio: 'pipe',
      })
      return true
    } catch (err) {
      console.warn('windows trust install failed', err.message)
      return false
    }
  }
  return false
}

File: ops/youtube-rtmp-relay/src-tauri/src/tls.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

pub struct CertPaths {
  pub key: PathBuf,
  pub cert: PathBuf,
}

pub fn cert_paths(userdata: &Path) -> Result<CertPaths, String> {
  let dir = userdata.join("tls");
  fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
  Ok(CertPaths {
    key: dir.join("server.key"),
    cert: dir.join("server.crt"),
  })
}

pub fn ensure_server_certs(userdata: &Path) -> Result<CertPaths, String> {
  let paths = cert_paths(userdata)?;
  if paths.key.exists() && paths.cert.exists() {
    return Ok(paths);
  }
  let status = Command::new("openssl")
    .args([
      "req",
      "-x509",
      "-newkey",
      "rsa:2048",
      "-sha256",
      "-days",
      "3650",
      "-nodes",
      "-keyout",
      paths.key.to_str().ok_or("key path")?,
      "-out",
      paths.cert.to_str().ok_or("cert path")?,
      "-subj",
      "/CN=127.0.0.1",
      "-addext",
      "subjectAltName=IP:127.0.0.1,DNS:localhost",
    ])
    .status()
    .map_err(|e| format!("openssl missing or failed: {e}"))?;
  if !status.success() {
    return Err("openssl cert generation failed".into());
  }
  Ok(paths)
}

pub fn install_trust(cert: &Path) -> bool {
  #[cfg(target_os = "macos")]
  {
    let home = std::env::var("HOME").unwrap_or_default();
    let keychain = format!("{home}/Library/Keychains/login.keychain-db");
    Command::new("security")
      .args([
        "add-trusted-cert",
        "-d",
        "-r",
        "trustRoot",
        "-k",
        &keychain,
        cert.to_str().unwrap_or(""),
      ])
      .status()
      .map(|s| s.success())
      .unwrap_or(false)
  }
  #[cfg(target_os = "windows")]
  {
    Command::new("certutil")
      .args(["-addstore", "-user", "Root", cert.to_str().unwrap_or("")])
      .status()
      .map(|s| s.success())
      .unwrap_or(false)
  }
  #[cfg(not(any(target_os = "macos", target_os = "windows")))]
  {
    let _ = cert;
    false
  }
}
